controllers/deployment.go

```diff
@@ -343,9 +356,38 @@ func buildDeployment(master *v1alpha1.CDAPMaster, name string, services ServiceG
 	return spec, nil
 }
 
+func mergeEnvVars(masterSpec *v1alpha1.CDAPMasterSpec, serviceSpec *v1alpha1.CDAPServiceSpec) ([]corev1.EnvVar, error) {
+	// Merge master and service-specific environment variables
+	seenEnvVars := make(map[string]bool)
+	env := make([]corev1.EnvVar, len(masterSpec.Env))
+	copy(env, masterSpec.Env)
+	for _, envVar := range serviceSpec.Env {
+		if _, ok := seenEnvVars[envVar.Name]; ok {
+			return nil, fmt.Errorf("duplicate env var %q in service spec %q", envVar.Name, serviceSpec.Name)
+		}
+		seenEnvVars[envVar.Name] = true
+		overwritten := false
+		for ind, masterEnvVar := range env {
+			if envVar.Name == masterEnvVar.Name {
+				// Check if the env var key is duplicated in the master spec
+				if overwritten {
+					return nil, fmt.Errorf("duplicate env var %q in master spec", envVar.Name)
+				}
+				// Overwrite any conflicting environment variables with the service-specific values
+				env[ind] = envVar
+				overwritten = true
+			}
+		}
+		if !overwritten {
+			env = append(env, envVar)
+		}
+	}
```

Comment:
This loops seem to be doing a couple of things which we should separate out to make it easier to understand:
1) Detecting duplicate env vars in MasterSpec and Service Spec. This logic can be put in a separate function which can be called for both specs.
2) Merging two lists while giving priority to service spec when the same env var is present in both specs. 

For 2 we can create an empty map[string]string or map[string]corev1.envvar and do the following:
1) Insert all env vars from master spec as key value pairs with var.Name as the key.
2) Repeat step 1 for service spec, this should overwrite same keys inserted in step 1.
3) Convert this map into a list of env vars.

This should make the code easier to follow.

Reply:
Another way which generalises both steps from previous comment into 1 is to create a helper function:
```
func insertIntoMap(m map[string]string, vals []corev1.envVar, allowOverwrite) err {
  // insert vals into m
  // return error if allowOverwrite is false and key already exists in m
}
```
With this, we can simply this loop into the following calls:
1) check insertIntoMap(make(map[string]string, masterSpec.env, false) == nil // no duplicate keys
2) check insertIntoMap(make(map[string]string, serviceSpec.env, false) == nil // no duplicate keys
3) 
```
m = make(map[string]string
insertIntoMap(m, masterSpec.env, true)
insertIntoMap(m, serviceSpec.env, true)
```
4) Convert m to an array and return.

Reply:
Fixed, I used a double map to do this, should be more clear now. Thanks!